### ISSUE TITLE
Formatting

### ISSUE THREAD

**Issue by bohlender** [bohlender/vim-z3-smt2#2]
I think it would be nice to add some auto-formatting capabilities. Currently, when writing or debugging SMT instances, I often format them manually **to review the structure** of each expression and identify potential errors.

To illustrate the kind of formatting I have in mind, consider the following [De Morgan example](https://rise4fun.com/Z3/tutorial/guide):
```SMT
(declare-const a Bool)
(declare-const b Bool)
(define-fun demorgan () Bool
    (= (and a b) (not (or (not a) (not b)))))
(assert (not demorgan))
(check-sat)
```
Even in this simple case the structure of `(= (and a b) (not (or (not a) (not b)))))` is not immediately clear and could use some reformatting for closer inspection.

I suggest that pressing `<localleader>f` should yield the following:
```SMT
(declare-const a Bool)
(declare-const b Bool)
(define-fun demorgan () Bool
    (=
        (and a b)
        (not
            (or
                (not a)
                (not b))))) ; Not sure about the parens -- maybe on their own lines?
(assert (not demorgan))
(check-sat)
```
Any opinions on that? Maybe there even is an easy way to achieve this already. It's not a pressing issue for sure, but I guess something along those lines would be nice to have.

**Comment by phlo** [comment]
Excellent idea, this would be really nice to have!

I've already been looking for a simple solution every now and then, but couldn't find anything feasible yet. Some Lisp auto-formatters exist, but the ones I tried would require quite some modifications for being useful.

From my perspective, the simple style you proposed already looks quite intuitive and would be easy to implement (I guess). Regarding closing parenthesis, I'd keep them on the final line and not waste to much vertical display space.

Unfortunately, I don't have much time right now to look into it in great detail, but will help as much as I can!

**Comment by bohlender** [comment]
Glad to hear that there's some interest.

> Unfortunately, I don't have much time right now to look into it in great detail, but will help as much as I can!

The same is true for me. However, lets try to get this implemented at some point -- I've put it on my todo-list.

One aspect that needs discussion is the general approach to implementing this, in particular decide on:

1. keeping the plugin **vimscript-only**, or
1. writing/reusing some **external auto-formatter**?

While writing vimscript would be tedious, this approach would not entail any additional dependencies. In contrast, writing or reusing an external tool will likely be easier, more configurable and less error-prone, but require additional tooling to be installed on the users' machines.

_Note: There might also be some middle ground -- extending some vimscript-only formatter -- but I've not looked into this yet._

@phlo: Since you've already looked into reutilising Lisp-formatters, I assume you find the latter approach more reasonable? Assuming there were a `smt-format` tool, would it be acceptable to require users to install/compile the external formatter themselves, or should it be bundled with the plugin somehow?

**Comment by krobelus** [comment]
I use this script for formatting SMT code. The only thing that's annoying is that `pyparsing` is extremely slow, it takes 2 seconds to format 1000 lines..

<details>

```python
#!/usr/bin/env python3

from pyparsing import (
    ZeroOrMore,
    Group,
    Optional,
    restOfLine,
    nestedExpr,
    quotedString,
    Forward,
    ParseResults,
)
from typing import Tuple
import re
import sys
from os.path import basename

COLUMN_LIMIT = 80
INDENT = " " * 2

USAGE = f"""\
usage: {basename(sys.argv[0])} < input.lisp

Pretty-print a balanced set of parentheses.

Short expressions (< {COLUMN_LIMIT} characters) are printed inline.
Larger expressions are broken up in lines and aligned.
"""

def format_lisp(input: str) -> str:
    parser = lisp_parser()
    paragraphs = re.split(r'\n{2,}', input)
    return "\n".join(format_terms(parser.parseString(p)) for p in paragraphs)

def lisp_parser() -> Forward:
    comment = Group(";" + Optional(restOfLine))
    return ZeroOrMore(nestedExpr(ignoreExpr=(quotedString | comment)) ^ comment)

def format_terms(xs: ParseResults) -> str:
    return "".join(format_term(x, 0, False) + "\n" for x in xs)

def iscomment(xs: ParseResults) -> bool:
    return len(xs) == 2 and xs[0] == ";"

def isatom(xs: ParseResults) -> bool:
    return isinstance(xs, str)

def format_term(xs: ParseResults, level: int, first: bool) -> str:
    # Insert comments at the current indentation level.
    if iscomment(xs):
        return "".join(xs)
    # Atoms are easy, always print them as-is.
    if isatom(xs):
        return xs
    ok, oneline = format_term_oneline(xs)
    if ok and len(oneline) < COLUMN_LIMIT:  # Small terms on one line
        return oneline
    # Long expression, break lines and align subexpressions on the same level.
    s = "("
    if not iscomment(xs[0]) and not isatom(xs[0]):
        s += "\n"
    # General lists with line breaks after each element.
    for i in range(len(xs)):
        x = xs[i]
        s += format_term(x, level + 1, i == 0)
        if i != len(xs) - 1:
            s += "\n"
    s += ")"
    indented = ""
    first = True
    for line in s.splitlines(keepends=True):
        if not first:
            indented += INDENT
        indented += line
        first = False
    return indented

def format_term_oneline(xs: ParseResults) -> Tuple[bool, str]:
    if iscomment(xs):
        return False, ""
    if isatom(xs):
        return True, xs
    terms = []
    for x in xs:
        ok, s = format_term_oneline(x)
        if not ok:
            return False, ""
        terms += [s]
    return True, "(" + " ".join(terms) + ")"

if __name__ == "__main__":
    if len(sys.argv) > 1:
        print(USAGE)
        sys.exit(1)
    else:
        print(format_lisp(sys.stdin.read()), end="")

TESTDATA = (
    """
(simple x)

(list (list (fun) (number 0.0.0.0:5900) atom))

;; comment

(list
  ; Comments are aligned too!
  (string "string literal with("))

(let
  ; List of lists.
  ((x y) (y x)))

; Short expressions on one line
(let ((x y) (y x)))

; Longer expressions are broken up and aligned.
(assert
  (=
    x
    (Pointer
      true
      #x00000002
      (variant_node1 (Pointer true #x00000001 variant_leaf1)))))

(declare-datatypes
  ((Pointer 0) (Any 0))
  (
    ((Pointer (? Bool) (address (_ BitVec 32)) (* Any)))
    (
      (variant_leaf1)
      (variant_leaf2)
      (variant_node1 (node1.next Pointer))
      (variant_node2 (node2.next Pointer)))))
""".strip()
    + "\n"
)

def test_format_lisp() -> None:
    assert format_lisp(TESTDATA) == TESTDATA
```

</details>

**Comment by phlo** [comment]
Thank you very much for sharing this example! We'll look into it and might steal a line or two. :wink:

**Comment by bohlender** [comment]
Toying around with @krobelus code, I found that we can easily get rid of the `pyparsing` dependency by implementing some basic parser combinators:

<details><summary>Code</summary>

```python3
#!/usr/bin/env python3

import re
from typing import Tuple

######################################################################
# Generic parser builders
######################################################################
# Here, a parser is a function of the following form:
# f(input: str) -> (success: bool, output: str, value)

def regex(pat: str, func=lambda x: None):
    def f(s: str):
        m = re.compile(pat).match(s)
        if m:
            return True, s[m.end(0):], func(s[:m.end(0)])
        else:
            return False, s, None
    return f

def seq(*parsers):
    def f(s: str):
        cur_s = s
        values = list()
        for parser in parsers:
            succ, cur_s, value = parser(cur_s)
            if not succ:
                return False, s, None
            if value:
                values.append(value)
        return True, cur_s, values
    return f

def choice(*parsers):
    def f(s: str):
        cur_s = s
        for parser in parsers:
            succ, cur_s, value = parser(cur_s)
            if succ:
                return True, cur_s, value
        return False, s, None
    return f

def zeroOrMore(parser):
    def f(s: str):
        cur_s = s
        values = list()
        while True:
            succ, cur_s, value = parser(cur_s)
            if not succ:
                break
            values.append(value)
        return True, cur_s, values
    return f

def oneOrMore(parser):
    def f(s: str):
        if not parser(s)[0]:
            return False, s, None
        return zeroOrMore(parser)(s)
    return f

######################################################################
# Specific parser builders
######################################################################
# Paragraph ::= (Comment | SExpr)+
# SExpr     ::= '(' Expr* ')'
# Expr      ::= Comment | SExpr | atom

def lparen():
    return regex(r'^\s*\(')

def rparen():
    return regex(r'^\s*\)')

def paragraph():
    def f(s: str):
        parser = oneOrMore(choice(comment(), sexpr()))
        return parser(s)
    return f

def comment():
    def f(s: str):
        parser = regex(r'^\s*;.*', lambda x: [';', x.strip()[1:]])
        return parser(s)
    return f

def sexpr():
    def f(s: str):
        parser = seq(lparen(), zeroOrMore(expr()), rparen())
        succ, cur_s, value = parser(s)
        # Do not wrap Expr* value in another list
        if succ and value:
            return succ, cur_s, value[0]
        return succ, cur_s, value
    return f

def expr():
    def f(s: str):
        parser = choice(comment(), sexpr(), atom())
        return parser(s)
    return f

def atom():
    def f(s: str):
        quoted = regex(r'^\s*".*"', lambda x: x.strip())
        # identifier = all but whitespace and ( or )
        id = regex(r'^\s*[^\s\(\)]+', lambda x: x.strip())
        return choice(quoted, id)(s)
    return f

######################################################################
# Formatter & test (effectively copy-paste from @krobelus)
######################################################################

COLUMN_LIMIT = 80
INDENT = " " * 2

def format_lisp(input: str) -> str:
    parser = paragraph()
    paragraphs = re.split(r'\n{2,}', input)
    return "\n".join(format_terms(parser(p)[2]) for p in paragraphs)

def format_terms(xs) -> str:
    return "".join(format_term(x, 0, False) + "\n" for x in xs)

def iscomment(xs) -> bool:
    return len(xs) == 2 and xs[0] == ";"

def isatom(xs) -> bool:
    return isinstance(xs, str)

def format_term(xs, level: int, first: bool) -> str:
    # Insert comments at the current indentation level.
    if iscomment(xs):
        return "".join(xs)
    # Atoms are easy, always print them as-is.
    if isatom(xs):
        return xs
    ok, oneline = format_term_oneline(xs)
    if ok and len(oneline) < COLUMN_LIMIT:  # Small terms on one line
        return oneline
    # Long expression, break lines and align subexpressions on the same level.
    s = "("
    if not iscomment(xs[0]) and not isatom(xs[0]):
        s += "\n"
    # General lists with line breaks after each element.
    for i in range(len(xs)):
        x = xs[i]
        s += format_term(x, level + 1, i == 0)
        if i != len(xs) - 1:
            s += "\n"
    s += ")"
    indented = ""
    first = True
    for line in s.splitlines(keepends=True):
        if not first:
            indented += INDENT
        indented += line
        first = False
    return indented

def format_term_oneline(xs) -> Tuple[bool, str]:
    if iscomment(xs):
        return False, ""
    if isatom(xs):
        return True, xs
    terms = []
    for x in xs:
        ok, s = format_term_oneline(x)
        if not ok:
            return False, ""
        terms += [s]
    return True, "(" + " ".join(terms) + ")"


TESTDATA = (
    """
(simple x)

(list (list (fun) (number 0.0.0.0:5900) atom))

;; comment

(list
  ; Comments are aligned too!
  (string "string literal with("))

(let
  ; List of lists.
  ((x y) (y x)))

; Short expressions on one line
(let ((x y) (y x)))

; Longer expressions are broken up and aligned.
(assert
  (=
    x
    (Pointer
      true
      #x00000002
      (variant_node1 (Pointer true #x00000001 variant_leaf1)))))

(declare-datatypes
  ((Pointer 0) (Any 0))
  (
    ((Pointer (? Bool) (address (_ BitVec 32)) (* Any)))
    (
      (variant_leaf1)
      (variant_leaf2)
      (variant_node1 (node1.next Pointer))
      (variant_node2 (node2.next Pointer)))))
""".strip()
    + "\n"
)

if __name__ == "__main__":
    assert format_lisp(TESTDATA) == TESTDATA
```
</details>

Now that we have a dependency-free reference implementation (which mostly relies on regular expressions), we should be able to add auto-formatting by "translating" this to vimscript.

**edit**: Fixed regexes not starting with `^` (@krobelus: don't miss the fix if you use this)

**Comment by krobelus** [comment]
This is amazing, it is 20x faster :smile: thanks so much!
I usually don't use Vim, so I'll just integrate this in my editor as external script, but I'll share any improvements.
The nice thing is it's not specific to SMT at all, it works for any lisp.

**Comment by bohlender** [comment]
I've ported the above approach to vimscript on a [feature branch](https://github.com/bohlender/vim-smt2/tree/feature/lisp-autoformat) -- give it a try @phlo.

I found it working well for small examples but the feature still needs some work. There are some TODOs left in the code but most importantly we have to improve the performance. Right now, even [a file of 160KB](https://github.com/sosy-lab/sv-benchmarks/blob/master/clauses/LIA/qarmc/amotsa.smt2) takes ~15s to parse (the above python code takes <1s). Caching would be an obvious first step.

**Comment by phlo** [comment]
Awesome, great work @bohlender! :1st_place_medal: :nerd_face:

It indeed works pretty nicely for paragraphs, which is the main use-case I guess. A quick review just showed one tiny "bug", where a *short expression* may exceed the overall `g:smt2_formatter_short_length` limit due to the prepended indentation string not being taken into account when calculating the resulting line's length. I think this can be fixed by simply adding `len(indent_str)` or `len(g:smt2_formatter_indent_str) * a:indent` to the following condition:
https://github.com/bohlender/vim-smt2/blob/33629fbc7fa4a938e5ad95a32a18afe714a2713d/autoload/smt2/formatter.vim#L57

Regarding performance: I could not even format a ~178K file, as it failed with `E132: Function call depth is higher than 'maxfuncdepth'`. So, why not just blame recursion? :thinking:

Unfortunately, I'm still desperately trying to meet a deadline and will have to delay further efforts for at least another couple of weeks. However, your port really is a nice starting point and I hope that I'll be able to help rather sooner than later!

Also, thanks again to @krobelus for sparking the flame! Greets to Markus and Evelyn by the way ... :wink:

**Comment by bohlender** [comment]
> A quick review just showed one tiny "bug", where a short expression may exceed the overall `g:smt2_formatter_short_length` limit due to the prepended indentation string not being taken into account when calculating the resulting line's length.

I intended `g:smt2_formatter_short_length` to limit the length of what is considered as a "short expression" -- not the length of the line before breaking. That's also why I renamed the variable from `COLUMN_LIMIT`. Maybe the default of 80 chars was still misleading :sweat_smile: My intuition is that indenting an expression does not make it less readable, however, we can refer to the line length instead of the expression length if you think that's more useful.

Regarding performance, I'm not sure whether recursion is the core problem. It is *a* problem for sure but you can avoid `E132` via `set maxfuncdepth=1000000`. It seems like calling user defined Vim functions is expensive in general. For example, the following Vim script takes 2.7s to execute
```vim
function! Plus(lhs, rhs)
    return a:lhs + a:rhs
endfunction

function! Iterate(num)
    let res = 0
    for _ in range(1, a:num)
        let res = Plus(res, 1)
    endfor
    return res
endfunction

let start = reltime()
call Iterate(1000000)
echom 'Iterative:' . reltimestr(reltime(start)) . 's'
```
but equivalent code is negligibly fast even in Python.

Unfortunately, with the current approach, parsing `(declare-const a Bool)` already involves 145 parser function calls. So it's not surprising that real-world SMT files take a while to parse. I will evaluate a non-recursive approach but I expect it to not be significantly faster. So far I found that inlining and avoiding closures can easily improve the performance by about 25% but that just makes the code less readable and is still uncomfortably slow.

I'm not sure what other options we have to make this work fast, besides trying [Vim9 script](https://vimhelp.org/vim9.txt.html#Vim9-script) or Vim's [Python interface](https://vimhelp.org/if_pyth.txt.html#if_pyth.txt). I hope we're just missing something obvious.

> Unfortunately, I'm still desperately trying to meet a deadline and will have to delay further efforts for at least another couple of weeks. However, your port really is a nice starting point and I hope that I'll be able to help rather sooner than later!

Sure -- we're not in a hurry :wink:

**Comment by krobelus** [comment]
I've pushed my script to https://github.com/symflower/smtfmt, with the enhancements from @bohlender, I hope that's okay. I haven't made any logic changes.
I like an external tool here because it's more flexible than a Vimscript implementation.

**Comment by bohlender** [comment]
Just to keep you updated: In the meantime I've experimented with various implementations (recursive & iterative) in `vimscript`, `vim9script` and `D` (to have a native reference implementation).

It turns out that `vimscript` won't get us far since the overhead of interpreting is just too high. With `vim9script` functions do get compiled and don´t have to be re-interpreted over and over again. This improves the performance by orders of magnitude but is still orders of magnitude slower than `D`. The major problem with our current parsing approach is that, [in Vim, accessing a string at some index i has complexity O(i)](https://groups.google.com/g/vim_dev/c/037sOxXDcqg/m/0qXI6nw7AwAJ) -- not O(1) -- but we rely on this operation when using `matchend`, `matchstrpos` or just `slice` the input string.

With that in mind, it is clear that the approach needs to be restructured a bit. I'm thinking of sticking to `vim9script` but prepending a standard scanner/tokenizer pass so there's no need to index into the input string anymore.

**Comment by bohlender** [comment]
I've prepended a scanner/tokenizer to the parser in [feature/lisp-autoformat](https://github.com/bohlender/vim-smt2/tree/feature/lisp-autoformat). All the parsing & formatting functionality now relies on `vim9script` (the used subset should be included in Vim >= v8.2).

Give it a try @phlo. I find the scanning & parsing of this revised implementation to be even faster than our initial experiments with `Python`  -- still slower than native code though. However, it looks like there's unnecessary/expensive string-copying happening in the formatter so I've [uncommented it](https://github.com/bohlender/vim-smt2/blob/382b8ea3352701b8b480a01ec9f506729f31b3f6/ftplugin/smt2.vim#L9) for now and am printing the parsing result / the AST instead.

At this point it should be possible to work on improving the formatter and remaining TODOs in the scanner & parser independently.

**Comment by phlo** [comment]
Looks tremendous @bohlender!

Unfortunately, I ran into some troubles during my trials. After explicitly setting the `debug` variable's type in [parser.vim](https://github.com/bohlender/vim-smt2/blob/feature/lisp-autoformat/autoload/smt2/parser.vim) due to an `E1091: type missing for debug  = true` error, I now get `E1024: need a List to iterate over` while trying to print the AST in [line 37](https://github.com/bohlender/vim-smt2/blob/382b8ea3352701b8b480a01ec9f506729f31b3f6/autoload/smt2/parser.vim#L37) and was not yet able to find out why. I'm using v8.2.0814, which should support the required subset. Do you have any idea what might be the cause?

**Comment by bohlender** [comment]
> Unfortunately, I ran into some troubles during my trials. After explicitly setting the `debug` variable's type in [parser.vim](https://github.com/bohlender/vim-smt2/blob/feature/lisp-autoformat/autoload/smt2/parser.vim) due to an `E1091: type missing for debug = true` error, I now get `E1024: need a List to iterate over` while trying to print the AST in [line 37](https://github.com/bohlender/vim-smt2/blob/382b8ea3352701b8b480a01ec9f506729f31b3f6/autoload/smt2/parser.vim#L37) and was not yet able to find out why. I'm using v8.2.0814, which should support the required subset. Do you have any idea what might be the cause?

I could reproduce your issue with the suggested Vim version. The automatic type inference is a minor thing and easily fixed (as you did) but the latter issue seems to be due to `vim9script` being pretty buggy in that version -- I couldn't even run the scanner without segfaults. We should probably have expected this since `vim9script` is largely still under development.

I'm on patch version 2845 and `vim9script` seems reasonably stable at this point. However,  I'm not sure we can expect users to have a recent version installed :thinking:

Guess I'll tackle the weirdly slow formatter first. We'll see what to do about `vim9script` later.

**Comment by bohlender** [comment]
#### Minor update
The recent changes also enable auto-formatting of a solver's output:
[![asciicast](https://asciinema.org/a/417084.svg)](https://asciinema.org/a/417084)

I'm happy that this now works out of the box -- it bugged me in the first version.

**Comment by phlo** [comment]
> The recent changes also enable auto-formatting of a solver's output.

Neato! :heart_eyes:

> The latter issue seems to be due to `vim9script` being pretty buggy in that version. ... However, I'm not sure we can expect users to have a recent version installed. :thinking:

Oh really!? What a bummer! Currently, v8.2.0814 is the only version in Gentoo's portage tree, besides the bleeding edge. I did a quick survey to asses the situation. Below you find a table containing the current Vim versions of all major distributions I could think of.

| Distro | Version |
| --- | --- |
| SUSE Enterprise (15.3) | 8.0.1568 |
| openSUSE (15.2) | 8.0.1568 |
| CentOS (8.3) | 8.0.1763 |
| RHEL (8.3) | 8.0.1763 |
| Debian (stable) | 8.1.0875 |
| Ubuntu (20.04 LTS) | 8.1.2269 |
| Gentoo | 8.2.0814 |
| Alpine (3.13) | 8.2.2320 |
| Debian (testing) | 8.2.2434 |
| Ubuntu (21.04) | 8.2.2434 |
| openSUSE (Tumbleweed) | 8.2.2850 |
| Fedora (34) | 8.2.2879 |
| Arch | 8.2.2891 |
| Slackware | 8.2.2891 |

Looks like we actually can't expect users to have a proper version. I really like your `vim9script` approach, so let's hope this will be resolved in future releases. :worried:

**Comment by bohlender** [comment]
I've resolved all the important TODOs. Now everything runs pretty much as expected and (except for trying out non-recursive formatting) I don't see too much room for improvement. However, before we can merge this into `master`, we should decide how to proceed regarding the use of `vim9script`.

Thanks for the above list of vim versions in the various distributions. Using the [AppImage releases of vim](https://github.com/vim/vim-appimage/releases) I've found that the `vim9script` functionality we use is buggy until [8.2.2725](https://github.com/vim/vim-appimage/releases/tag/v8.2.2725). So, according to your list, the formatting will currently not work for most distros.

Which options do we have?
1. Since we only use `vim9script` for auto-formatting, merging it into `master` will not break existing functionality, e.g. invoking a solver will still work with an "old" Vim. However, formatting won't work for users whose Vim version is below 8.2.2725 -- I've added a corresponding error message.
1. We can provide an implementation in "regular" Vimscript for users of older Vim versions and switch depending on the major/minor version. However, without `vim9script` the performance is so bad (except for small files) that it's hardly worth having a non-`vim9script` implementation. Furthermore, it's tedious to maintain two implementations and keep them in sync.
1. Vimscript is known to be slow, so -- aside from `vim9script` -- Vim also supports calling `Python` or `Lua` code if higher performance is needed. However, I think that requiring another interpreter to be installed, and Vim to be compiled with [the corresponding option](https://vimhelp.org/if_pyth.txt.html#if_pyth.txt), is a worse dependency than just requiring a recent (>= 8.2.2725) Vim version.

Despite the fact that most distros are stuck with "older" Vim versions, I'm inclined to go with option 1 since formatting is essentially an optional feature which will eventually become available to everyone -- once the distros' packages get updated. I'm open for discussion and interested in other opinions though. Maybe there's an option/idea that I've missed.

**Comment by phlo** [comment]
Awesome, thanks again for your efforts!

I totally agree, since formatting is intended to be an optional feature I'd also go with option 1. Introducing additional dependencies and reimplementing the `vim9script` approach would be an overkill!

So from my side there's nothing speaking against merging [`feature/lisp-autoformat`](https://github.com/bohlender/vim-smt2/tree/feature/lisp-autoformat) into [`master`](https://github.com/bohlender/vim-smt2/tree/master)! :grinning:
